File: content/01-arrays-and-hashing/longest_consecutive_subsequence.py
"""
Problem: Longest Consecutive Subsequence
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/longest-consecutive-subsequence/0
"""

def longest_consecutive(arr: list[int]) -> int:
    """
    Time Complexity: O(N)
    Space Complexity: O(N)
    """
    num_set = set(arr)
    longest_streak = 0
    
    for num in num_set:
        # Check if it's the start of a sequence
        if num - 1 not in num_set:
            current_num = num
            current_streak = 1
            
            while current_num + 1 in num_set:
                current_num += 1
                current_streak += 1
                
            longest_streak = max(longest_streak, current_streak)
            
    return longest_streak

if __name__ == "__main__":
    assert longest_consecutive([2, 6, 1, 9, 4, 5, 3]) == 6
    assert longest_consecutive([1, 9, 3, 10, 4, 20, 2]) == 4
    assert longest_consecutive([]) == 0
    print("Python: longest_consecutive passed tests!")

File: content/01-arrays-and-hashing/longest_consecutive_subsequence.js
/**
 * Problem: Longest Consecutive Subsequence
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/longest-consecutive-subsequence/0
 */

function longestConsecutive(arr) {
    /**
     * Time Complexity: O(N)
     * Space Complexity: O(N)
     */
    const numSet = new Set(arr);
    let longestStreak = 0;
    
    for (const num of numSet) {
        if (!numSet.has(num - 1)) {
            let currentNum = num;
            let currentStreak = 1;
            
            while (numSet.has(currentNum + 1)) {
                currentNum++;
                currentStreak++;
            }
            
            longestStreak = Math.max(longestStreak, currentStreak);
        }
    }
    
    return longestStreak;
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(longestConsecutive([2, 6, 1, 9, 4, 5, 3]), 6);
    assert.strictEqual(longestConsecutive([1, 9, 3, 10, 4, 20, 2]), 4);
    assert.strictEqual(longestConsecutive([]), 0);
    console.log("JS: longestConsecutive passed tests!");
}
